[vcpkg/scripts/pkgconfig] rewrite pkg-config check for libraries

Neumann-A opened the pull request:
closes #11103

- [x] Testing with X windows PR (Linux)
- [x] Testing with X windows PR (Windows) (ran until xlib and than run into other problems)

Neumann-A committed:
[vcpkg/scripts/pkgconfig] rewrite pkgconfig check for libraries.

Neumann-A committed:
[vcpkg/scripts/pkgconfig] fix more details.

Neumann-A committed:
[vcpkg/scripts/pkgconfig] smaller improvements

JackBoosY commented:
Maybe we should choose one port to test this PR?

Neumann-A committed:
ws change to check functionallity of script for testing.

JackBoosY commented on ports/x264/portfile.cmake:
```diff
@@ -1,6 +1,6 @@
 set(X264_VERSION 157)
 
-vcpkg_fail_port_install(ON_TARGET "OSX") 
+vcpkg_fail_port_install(ON_TARGET "OSX")
```
Please update the version info. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).

Neumann-A committed:
[vcpkg/scripts/pkgconfig] update to latest test version

Neumann-A committed:
Revert "ws change to check functionallity of script for testing."

This reverts commit 26269e9e776009230777d9390465715f2ebac465.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_pkg_config_to_fixup_pkgconfig

Neumann-A committed:
[nettle] bump control for vcpkg_fixup_pkgconfig test

Neumann-A commented:
@JackBoosY: osx has no pkg-config installed?

Neumann-A committed:
[gmp] rerun CI to create new *.pc files

Neumann-A committed:
[gmp/nettle] enable port debug

Neumann-A committed:
[vcpkg/script/pkgconfig] fix path conversion regex for CI

JackBoosY commented:
Yes, please install `brew install pkgconfig` here:
https://github.com/microsoft/vcpkg/blob/d9b4acf02d8c784927f89d23edb677408ec428af/scripts/azure-pipelines/osx/azure-pipelines.yml#L25

Neumann-A committed:
[CI|OSX] install pkg-config on osx

Neumann-A committed:
[gmp/nettle] disable port debug

Neumann-A commented:
@JackBoosY: Done

JackBoosY commented:
Looks good in my side.

Neumann-A committed:
[vcpkg/scripts/pkgconfig] update to latest version of x windows pr

Neumann-A committed:
[zlib] add pkgconfig file

Neumann-A committed:
[zstd] add fixup pkgconfig

Neumann-A committed:
[libpng] add pc file installation

Neumann-A committed:
[bzip2] install pc file and leave a TODO for somebody else ;)

Neumann-A committed:
[bzip2] bump control

Neumann-A commented:
@JackBoosY: Sorry for adding more but I found a few flaws by running the x window pr a bit further. I also added a few *.pc files to basic ports for simple testing.

BillyONeal commented:
No changes to vm provisioning scripts here, dropping vm-update-request.

Neumann-A commented:
Hmm osx ci is not finding pkg-config again although the brew script claims it has been installed. Don't know what to do about that. Except for th fltk regression which seems to be an error in the baseline the rest are download errors.

Neumann-A committed:
[vcpkg/scripts/pkgconfig] skip checks if pkgconfig cannot be found.
- small bugfix in ignore flags regex

Neumann-A committed:
retry zstd and zlib

Neumann-A committed:
[libpng] retry by adding dl and m to system libraries for osx

JackBoosY commented:
Looks good now.

Neumann-A committed:
[libpng] add missing vcpkg_fixup_pkgconfig()

Neumann-A committed:
[vcpkg/scripts/pkgconfig] unset var if found. 

fixes checks for other configurations which will be always succesful otherwise

Neumann-A committed:
[libpng] bump control

Neumann-A committed:
[libpng] enable port debug to debug osx regression

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_pkg_config_to_fixup_pkgconfig

# Conflicts:
#	scripts/azure-pipelines/osx/azure-pipelines.yml

Neumann-A committed:
ws change to retrigger osx ci and hopeing that it works

JackBoosY commented:
/azp run

Neumann-A committed:
fix typo

JackBoosY commented:
Can you merge the latest changes into this branch?

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_pkg_config_to_fixup_pkgconfig

JackBoosY committed:
Merge branch 'master' into use_pkg_config_to_fixup_pkgconfig

JackBoosY commented:
Coud you resolve these file conflicts?

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_pkg_config_to_fixup_pkgconfig

# Conflicts:
#	scripts/azure-pipelines/osx/azure-pipelines.yml
#	scripts/cmake/vcpkg_fixup_pkgconfig.cmake

JackBoosY commented:
@BillyONeal This PR is important, we need to merge this first.

JackBoosY commented:
```
Building package libpng[core]:x64-osx...
...
-- Fixing pkgconfig
-- Fixing pkgconfig - release
-- Checking file: /Users/vagrant/Data/packages/libpng_x64-osx/lib/pkgconfig/libpng.pc
-- Checking file: /Users/vagrant/Data/packages/libpng_x64-osx/lib/pkgconfig/libpng16.pc
-- CHECK_LIB_m:CHECK_LIB_m-NOTFOUND
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:217 (message):
  Library "m" was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  correct the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:285 (vcpkg_fixup_pkgconfig_check_files)
  ports/libpng/portfile.cmake:117 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:73 (include)


Error: Building package libpng:x64-osx failed with: BUILD_FAILED
Elapsed time for package libpng:x64-osx: 8.888 s
```

Could you fix it?

Thanks.

Neumann-A committed:
[libpng] remove -lm flag on apple platforms

Neumann-A commented:
> Could you fix it?

Done. CI is testing the rest currently. Hopefully there is not more work to be done ;)

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_pkg_config_to_fixup_pkgconfig

# Conflicts:
#	scripts/cmake/vcpkg_fixup_pkgconfig.cmake

Neumann-A committed:
ws change to rerun some ports in ci for good measure after the merge with master

Neumann-A commented:
@JackBoosY: Please rerun CI for libpq:x86 and libnice. The libnice failure is probably unrelated the libpq:x86 error is just a download failure for winflexbison

Commit:
Re-trigger CI test

JackBoosY committed:
Merge branch 'master' into use_pkg_config_to_fixup_pkgconfig

Neumann-A commented:
@JackBoosY: needs a CI retrigger. Download errors in CI. Can this get merged any time soon?

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into use_pkg_config_to_fixup_pkgconfig

JackBoosY commented:
@ras0219  @ras0219-msft Could you merge this PR first?

ras0219-msft commented:
I'm going to try to merge https://github.com/microsoft/vcpkg/pull/12254 before this PR, which will ensure all consumers are re-tested.

Neumann-A committed:
[zstd] add pkgconfig

Neumann-A committed:
[zstd] correct debug link

Neumann-A committed:
Merge branch 'zstd_pkgconfig' of https://github.com/Neumann-A/vcpkg into zstd_pkgconfig

# Conflicts:
#	ports/zstd/portfile.cmake

ras0219-msft commented:
/azp run

since #12254 is merged

ras0219-msft commented:
/azp run

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_pkg_config_to_fixup_pkgconfig

Neumann-A commented:
currently checking why there is no regression in zstd. according to #12407 there should be a regression

Neumann-A committed:
[vcpkg/script/pkgconfig] - add more search suffixes for static libs
- add additional debug messages for debuging
- fix unsetting of cache variables hiding issues with debug libraries
- fixed pkgconfig fix in the debug case

Neumann-A committed:
Merge remote-tracking branch 'origin/zstd_pkgconfig' into use_pkg_config_to_fixup_pkgconfig

# Conflicts:
#	ports/zstd/CONTROL
#	ports/zstd/portfile.cmake

Neumann-A committed:
[pcre] add pthread dependency to pcre

Neumann-A committed:
[protobuf] fix pkgconfig

Neumann-A committed:
include quotes around the prefix in the case the prefix path was quoted (fixes some regressions)
add option NOT_STATIC_PKGCONFIG to run pkg-config without the --static argument

Neumann-A committed:
[ignition-msgs5] get rid of a dev warning

Neumann-A committed:
[apr] fix apr regression

Neumann-A committed:
[protobuf] fix protobuf regression

Neumann-A committed:
[sdl1] fix regression due to variable expansion of pkg-config. Use similar regex to the library removal for ignored flags.

Neumann-A committed:
[ignition] fix regression

Neumann-A commented:
@ras0219-msft: Should have 561287c trigger a full rebuild after 5ccacd4?

Please rerun failing CIs. This PR should be green now.

BillyONeal commented:
@Neumann-A Easiest way to force that is to push a merge with master; in this case there are merge conflicts

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into use_pkg_config_to_fixup_pkgconfig

# Conflicts:
#	ports/zlib/CONTROL
#	ports/zlib/portfile.cmake

Neumann-A committed:
remove changes of zlib portfile.

Neumann-A committed:
reset changes in osx ci pipeline

Neumann-A commented:
The ompl:x86-windows regressions seems unrelated. Worked locally and error in CI is:
```
CMake Error at tests/CMakeLists.txt:16 (configure_file):
  configure_file Problem configuring file
```

Commit:
add DISABLE_PARALLEL_CONFIGURE

JackBoosY commented:
@ras0219-msft @ras0219 ping for merge this PR.

ras0219-msft commented:
LGTM, thanks @Neumann-A for the continued awesome work on this!

I think that it will eventually make sense to move the "checking" parts of this into the vcpkg tool itself, to ensure it is always run on any `.pc` files left in the tree, but this is great for now.
